### PR TITLE
grlib-multi: implement SpaceWire driver

### PR DESCRIPTION
**Pull request by lukileczo**
JIRA: RTOS-709

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## How Has This Been Tested?

- [ ] Already covered by automatic testing.
- [x] New test added: https://github.com/phoenix-rtos/phoenix-rtos-tests/pull/307
- [ ] Tested by hand on: (list targets here).

## Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [x] All new and existing linter checks and tests passed.
- [x] My changes generate no new compilation warnings for any of the targets.

## Special treatment

- [ ] This PR needs additional PRs to work (list the PRs, preferably in merge-order).
- [ ] I will merge this PR by myself when appropriate.

### CODE REVIEW

File: multi/grlib-multi/spacewire.c [hunk 1]
```diff
@@ -0,0 +1,828 @@
...
+		dev->rxAcknowledged[i] = true;
+	}
+	dev->rxBuff = MAP_FAILED;
+	dev->txBuff = MAP_FAILED;
+	dev->rxDesc = MAP_FAILED;
+	dev->txDesc = MAP_FAILED;
+
+	uintptr_t base = (pbase & ~(_PAGE_SIZE - 1));
+	dev->vbase = mmap(
+		NULL, _PAGE_SIZE, PROT_WRITE | PROT_READ, MAP_DEVICE | MAP_PHYSMEM | MAP_ANONYMOUS, -1, (off_t)base);
```
Comment (agkaminski) [inline multi/grlib-multi/spacewire.c:701]:
```suggestion
	dev->vbase = mmap(NULL, _PAGE_SIZE, PROT_WRITE | PROT_READ, MAP_DEVICE | MAP_PHYSMEM | MAP_ANONYMOUS, -1, (off_t)base);
```

File: multi/grlib-multi/spacewire.h [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+void spw_handleMsg(msg_t *msg, int dev);
+
+
+int spw_createDevs(oid_t *oid);
+
+
+int spw_init(void);
+
+
+#endif
```
Comment (agkaminski) [inline multi/grlib-multi/spacewire.h:30]: No new line at end of the file

File: multi/grlib-multi/spacewire.c [hunk 3]
```diff
@@ -0,0 +1,828 @@
...
+			/* Wrap around */
+			desc->ctrl |= TX_DESC_WR;
+		}
+		desc->packetLen = packet.dataLen;
+
+		uint8_t hdrLen = packet.flags & TX_DESC_HDR_LEN;
+		volatile void *txBuff = dev->txBuff[dev->lastTxDesc];
+
+		memcpy((void *)txBuff, packet.hdr, hdrLen);
+		memcpy((void *)txBuff + hdrLen, packet.data, packet.dataLen);
```
Comment (agkaminski) [inline multi/grlib-multi/spacewire.c:424]: void * arithmetic
```suggestion
		memcpy((char *)txBuff + hdrLen, packet.data, packet.dataLen);
```

File: multi/grlib-multi/spacewire.c [hunk 4]
```diff
@@ -0,0 +1,828 @@
...
+
+		desc->hdrAddr = (uint8_t *)va2pa((void *)txBuff);
+		desc->dataAddr = (uint8_t *)va2pa((void *)txBuff + hdrLen);
```
Comment (agkaminski) [inline multi/grlib-multi/spacewire.c:427]:
```suggestion
		desc->hdrAddr = (uint8_t *)va2pa((void *)txBuff);
		desc->dataAddr = (uint8_t *)va2pa((char *)txBuff + hdrLen);
```

File: multi/grlib-multi/spacewire.c [hunk 5]
```diff
@@ -0,0 +1,828 @@
...
+		if (dev->lastTxDesc == SPW_TX_DESC_CNT - 1) {
+			/* Wrap around */
+			desc->ctrl |= TX_DESC_WR;
+		}
+		desc->packetLen = packet.dataLen;
+
+		uint8_t hdrLen = packet.flags & TX_DESC_HDR_LEN;
+		volatile void *txBuff = dev->txBuff[dev->lastTxDesc];
+
+		memcpy((void *)txBuff, packet.hdr, hdrLen);
```
Comment (agkaminski) [inline multi/grlib-multi/spacewire.c:423]: Not the best practice to memcpy to volatile buff, doesn't matter in this case I guess

File: multi/grlib-multi/spacewire.c [hunk 6]
```diff
@@ -0,0 +1,828 @@
...
+		if ((dev->txDesc[firstDesc].ctrl & TX_DESC_EN) == 0) {
+			dev->txWaited[firstDesc] = false;
+			size_t next = (firstDesc + 1) % SPW_TX_DESC_CNT;
+			if ((next == 0) && wrapped) {
+				wrapped = false;
+			}
+			firstDesc = next;
+		}
+		else {
+			(void)condWait(dev->cond, dev->txLock, 0);
```
Comment (agkaminski) [inline multi/grlib-multi/spacewire.c:379]: Is it ok to risk losing `txLock` here? Perhaps some other thread might seize the opportunity and start new `spw_transmitWait` on it's own.

File: multi/grlib-multi/spacewire.c [hunk 7]
```diff
@@ -0,0 +1,828 @@
...
+#include <sys/types.h>
+#include <sys/threads.h>
+#include <posix/utils.h>
+
+#if defined(__CPU_GR716)
+#include <phoenix/arch/gr716.h>
+#elif defined(__CPU_GR712RC)
+#include <phoenix/arch/gr712rc.h>
+#else
+#error "Unsupported target"
```
Comment (agkaminski) [inline multi/grlib-multi/spacewire.c:34]:
```suggestion
#error Unsupported target
```

File: multi/grlib-multi/grlib-multi.c [hunk 8]
```diff
@@ -301,10 +311,77 @@ static void uart_thread(void *arg)
...
+		while (msgRecv(multi_common.spwOid.port, &msg, &rid) < 0) {
+		}
+
+		switch (msg.type) {
+			case mtDevCtl:
+				spw_dispatchMsg(&msg);
+				break;
+
+			case mtGetAttr:
+			case mtSetAttr:
```
Comment (agkaminski) [inline multi/grlib-multi/grlib-multi.c:352]: `*Attr` stuff use different msg field

File: multi/grlib-multi/spacewire.c (hunk 5 above)
Comment (lukileczo) [reply multi/grlib-multi/spacewire.c:423]: I agree, here volatile doesn't apply just yet

File: multi/grlib-multi/spacewire.c (hunk 6 above)
Comment (lukileczo) [reply multi/grlib-multi/spacewire.c:379]: Ouch, not really, deadlock could happen now

File: multi/grlib-multi/grlib-multi.c [hunk 9]
```diff
@@ -301,10 +311,78 @@ static void uart_thread(void *arg)
...
+				msg.o.attr.err = -ENOSYS;
```
Comment (agkaminski) [inline multi/grlib-multi/grlib-multi.c:353]:
```suggestion
				msg.o.attr.err = -ENOSYS;
				break;
```

**Review by agkaminski — DISMISSED** [review]
LGTM

File: multi/grlib-multi/spacewire.c [hunk 10]
```diff
@@ -0,0 +1,829 @@
...
+ * 24:0 - packet length (valid after EN was set to 0, in bytes)
+ *
+ * Bits IE, WR, EN are set to 0 by the hardware after packet is received.
+ * When EN is 0, status bits and packet length are valid and can be read.
+ * CRC header calculation can only be used if packet is of type RMAP.
+ * CRC data calculation can be checked provided that sender has calculated
+ * CRC for the packet according to the RMAP standard.
+ */
+
+#define RX_DESC_TRUNC   (1 << 31)
```
Comment (kemonats) [inline multi/grlib-multi/spacewire.c:128]:
```suggestion
#define RX_DESC_TRUNC   (1U << 31)
```

File: multi/grlib-multi/spacewire.c [hunk 11]
```diff
@@ -0,0 +1,829 @@
...
+		/* Start transmission */
+		dev->vbase[DMA_CTRL] |= DMA_CTRL_TE;
+
+		dev->lastTxDesc = (dev->lastTxDesc + 1) % SPW_TX_DESC_CNT;
+	}
+
+	TRACE("Packets set up");
+
+	/* Wait for transmission to finish */
+	while ((firstDesc <= lastDesc) || (wrapped && (firstDesc > lastDesc))) {
```
Comment (kemonats) [inline multi/grlib-multi/spacewire.c:369]:
```suggestion
	while ((firstDesc <= lastDesc) || wrapped) {
```
`firstDesc > lastDesc` is always `true`.

File: multi/grlib-multi/spacewire.c [hunk 12]
```diff
@@ -0,0 +1,829 @@
...
+
+	size_t cnt = 0;
+	const size_t lastDesc = (firstDesc + nPackets) % SPW_RX_DESC_CNT;
+	bool wrapped = (lastDesc <= firstDesc);
+
+	TRACE("first: %u last: %u nPackets: %u", firstDesc, lastDesc, nPackets);
+
+	(void)mutexLock(dev->rxLock);
+
+	while ((firstDesc < lastDesc) || (wrapped && (firstDesc >= lastDesc))) {
```
Comment (kemonats) [inline multi/grlib-multi/spacewire.c:527]:
```suggestion
	while ((firstDesc < lastDesc) || wrapped) {
```
`firstDesc >= lastDesc` is always `true`.

**Review by agkaminski — APPROVED** [review]

**Pull request merged**